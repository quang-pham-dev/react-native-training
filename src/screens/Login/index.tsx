import React, {lazy, Suspense, useCallback, useLayoutEffect} from 'react'
import {Alert, TouchableOpacity} from 'react-native'

// Libs
import {s, vs} from 'react-native-size-matters/extend'

// Contexts
import {useAuthContext} from '@contexts/auth/AuthContext'
import {
  SIGN_IN,
  SIGN_IN_FAILED,
  SIGN_IN_SUCCESS,
} from '@contexts/auth/action/auth'

// Components
import LoadingIndicator from '@components/LoadingIndicator'
import {ErrorBoundary, ErrorMode} from '@components'

// Styles
import HeadingStyled from '@components/Heading/Heading.styles'
import IconStyled from '@components/Icon/Icon.styles'
import ViewStyled from '@components/View/View.styles'
import AnimatedKeyboard from '@components/AnimatedKeyboard'

// Navigator
import {NavigationPropsType} from '@navigators/app-navigator'

// Constants
import {AUTH_DATA, HEADING_TYPE} from '@constants'

// Themes
import {Icons, Metrics} from '@themes'

// Types
import {LoginCredentials} from '@model-types'

// Api
import {authService} from '@apis'

// Services
import {set} from '@services'

// Splitting lazy load component
export const LoginFormLazy = lazy(
  () => import('@screens/Login/components/LoginForm'),
)

type LoginProps = {
  navigation: NavigationPropsType
}

const Login = ({navigation}: LoginProps) => {
  const {dispatch} = useAuthContext()

  // handle action call api Login when user press Login button
  const handleSubmitButton = async (
    loginInfo: LoginCredentials,
  ): Promise<void> => {
    dispatch({type: SIGN_IN})

    const {username, password} = loginInfo

    try {
      const response = await authService.signIn(username, password)
      const dataStorage = JSON.stringify(response?.data)
      // set auth data info to local storage
      await set(AUTH_DATA, dataStorage)
      const user = response?.data?.user
      const access_token = response?.data
      if (user && access_token) {
        dispatch({
          type: SIGN_IN_SUCCESS,
          payload: {
            user,
            access_token,
          },
        })
      }
    } catch (error: any) {
      dispatch({type: SIGN_IN_FAILED, payload: error})
      error?.response?.data
        ? Alert.alert(error.response.data.message)
        : Alert.alert('Login failed', error.message)
    }
  }
  const handleBackArrow = useCallback(() => {
    navigation.goBack()
  }, [navigation])

  useLayoutEffect(() => {
    navigation.setOptions({
      headerBackVisible: false,
      headerLeft: () => (
        <TouchableOpacity onPress={handleBackArrow}>
          <IconStyled width={s(45)} height={s(45)} source={Icons.back} />
        </TouchableOpacity>
      ),
    })
  }, [handleBackArrow, navigation])

  return (
    <ErrorBoundary errorMode={ErrorMode.ALWAYS}>
      <>
        <AnimatedKeyboard>
          <>
            <ViewStyled.Custom flex={1} pTop={vs(105)}>
              <HeadingStyled textAlign="center" type={HEADING_TYPE.H1}>
                Welcome
              </HeadingStyled>
              <ViewStyled.Custom pTop={vs(Metrics.padding.tiny)}>
                <HeadingStyled textAlign="center" type={HEADING_TYPE.H4}>
                  Please enter your data to continue
                </HeadingStyled>
              </ViewStyled.Custom>
            </ViewStyled.Custom>
            {/* end header */}

            {/* Form */}
            <Suspense fallback={<LoadingIndicator />}>
              <ViewStyled.Custom flex={2} pTop={vs(30)}>
                <LoginFormLazy onSubmit={handleSubmitButton} />
              </ViewStyled.Custom>
            </Suspense>
            {/* end Form */}
          </>
        </AnimatedKeyboard>
      </>
    </ErrorBoundary>
  )
}

export default Login
