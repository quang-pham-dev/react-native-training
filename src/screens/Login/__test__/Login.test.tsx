import React from 'react'

// Libs
import {render} from '@testing-library/react-native'

// Screen
import Login from '../index'

jest.useFakeTimers()

const props: any = {
  navigation: {
    navigate: jest.fn(),
    setOptions: jest.fn(),
  },
}

describe('Login Screen', () => {
  test('Should render correctly', async () => {
    const component = render(<Login {...props} />)

    component.toJSON()

    expect(component).toMatchSnapshot()
  })
})
