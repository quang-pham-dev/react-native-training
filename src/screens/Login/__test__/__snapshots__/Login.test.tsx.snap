// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Screen Should render correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
  style={
    Array [
      Object {},
    ]
  }
>
  <View>
    <View
      animatedStyle={
        Object {
          "value": Object {
            "transform": Array [
              Object {
                "translateY": -0,
              },
            ],
          },
        }
      }
      collapsable={false}
      style={
        Object {
          "flex": 1,
          "transform": Array [
            Object {
              "translateY": -0,
            },
          ],
        }
      }
    >
      <View
        flex={1}
        pTop={105}
        style={
          Array [
            Object {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "paddingTop": 105,
              "width": "100%",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#1D1E20",
                "fontFamily": "Inter-SemiBold",
                "fontSize": 28,
                "lineHeight": 30,
                "textAlign": "center",
              },
            ]
          }
          textAlign="center"
          type="h1"
        >
          Welcome
        </Text>
        <View
          pTop={5}
          style={
            Array [
              Object {
                "paddingTop": 5,
                "width": "100%",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#8F959E",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 15,
                  "lineHeight": 16.5,
                  "textAlign": "center",
                },
              ]
            }
            textAlign="center"
            type="h4"
          >
            Please enter your data to continue
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "width": "100%",
            },
          ]
        }
      >
        <ActivityIndicator
          color="#FFFFFF"
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;
